logic_tree_tools.py

```diff
@@ -0,0 +1,972 @@
+"""
+Contains functions to modify logic trees.
+"""
+
+import copy
+from typing import Optional, Union
+
+import numpy as np
+from nzshm_model.logic_tree import (
+    GMCMLogicTree,
+    SourceLogicTree,
+)
+from nzshm_model.logic_tree.correlation import (
+    LogicTreeCorrelations,
+)
+
+from run_toshi_hazard_post_helper import CustomLogicTreePair
+
+
+def reduce_logic_tree_to_nth_highest_weighted_branch(
+    logic_tree: Union[SourceLogicTree, GMCMLogicTree], nth_highest: int
+) -> Union[SourceLogicTree, GMCMLogicTree]:
+    """
+    Reduce a logic tree to only the nth highest weighted branch in each branch set.
+    The highest weighted branch is the 1st highest weighted branch (nth_highest = 1).
+    The second-highest weighted branch is the 2nd highest weighted branch (nth_highest = 2) etc.
+
+    Parameters
+    ----------
+    logic_tree : SourceLogicTree or GMCMLogicTree
+        The logic tree to be modified.
+
+    nth_highest : int
+        The nth highest weighted branch to reduce the logic tree to. For example, if n = 1, the logic tree is reduced
+        to the first highest weighted branch. If n = 2, it is reduced to the second-highest weighted branch, etc.
+
+    Returns
+    -------
+    modified_logic_tree : SourceLogicTree or GMCMLogicTree
+        The logic tree after being reduced only consisting of the nth highest branch in each branch_set.
+
+    Raises
+    ------
+    ValueError
+        If the branches in the modified_logic_tree do not have valid weights.
+        That is its branches in each branch_set do not sum to 1.0.
```

Comment:
```suggestion
        If the branches in each branch set modified_logic_tree do not sum to 1.0.
```
Just describe the exact condition that produces the error rather than having an extra sentence.